R/read_write.R

```diff
@@ -158,10 +157,46 @@ h5check <- function(fname, group, name) {
   bs <- adf( target_id = main_est$target_id )
   bs$est_counts <- as.numeric(rhdf5::h5read(fname, paste0("bootstrap/bs", i)))
   bs$tpm <- counts_to_tpm(bs$est_counts, main_est$eff_len)
-
   bs
 }
 
+
+#'
+#' Computes the bootstrap summary statistics of each transcript in a kallisto
+#' object.
+#'
+#'
+#' @param num_transcripts the number of transcripts
+#' @param fname the file name for the HDF5 file
+#' @param num_bootstraps the number of bootstraps
+#' @param transform a function to apply to each bootstrap
+#' @param est_counts_sf a double to divide all the bootstraps by
+#' @return a \code{data.frame} containing with columns for the summary statistics
+#' and a row for each transcript
+read_bootstrap_statistics <- function(fname, num_bootstraps,
+  num_transcripts, est_count_sf, transform = function(x) log(x + 0.5)) {
+  bs_mat <- do.call(rbind, lapply(0:(num_bootstraps[1] - 1), function(i) {
+    transform(rhdf5::h5read(fname, paste0("bootstrap/bs", i)) / est_count_sf)
+  }))
+
+  bs_var <- apply(bs_mat, 2, var)
+}
+
+read_bootstrap_quant <- function(fname, num_bootstraps) {
+  bs <- do.call(cbind, lapply(0:(num_bootstraps[1] - 1), function(i) {
+    as.numeric(rhdf5::h5read(fname, paste0("bootstrap/bs", i)))
+  }))
+
+  eff_len <- rhdf5::h5read(fname, "aux/eff_lengths")
+
+  bs_quant_est_counts <- aperm(apply(bs, 1, quantile))
```

Comment:
I suspect part of the issue that you are having with speed is due to going across to mention 1 rather than dimension 2 like in `read_bootstrap_statistics`. This type of memory access in R is very slow due to the way that matrices are allocated. Basically, you are not being very cache friendly.